src/index.js

```diff
@@ -180,26 +223,62 @@ class Bitswap {
       blockListeners[cidStr] = (block) => {
         this.wm.cancelWants([cid])
         cleanupListener(cid)
-        callback(null, block)
+        retrieved.push(block)
+
+        if (retrieved.length === missing.length) {
+          finish(callback)
+        }
       }
 
-      this.notifications.once(unwantEvent, unwantListeners[cidStr])
-      this.notifications.once(blockEvent, blockListeners[cidStr])
+      this.notifications.once(
+        unwantEvent(cidStr),
+        unwantListeners[cidStr]
+      )
+      this.notifications.once(
+        blockEvent(cidStr),
+        blockListeners[cidStr]
+      )
     }
```

Comment:
Glad you agreed https://github.com/ipfs/js-ipfs-bitswap/pull/142/files/d48c20e091c90c1abb6f42862564ace978d09afc..f634c3a69304d0f3b3ba93480686272cca3ba1b7#r135027986